automata/schema_search_tools.py

```diff
@@ -0,0 +1,675 @@
+        ):  # check if the annihilation or generation rule is partially in the schemata. Sometimes it is a part of the rule with "#" in the middle and isn't on its own.
+            if (
+                rule[1] in ["0", 0]
+            ):  # if the output of the rule is 0, the corresponding maintenance rule output must also be 0
+                # modify the middle value of schemata(wildcard_rule) to "0"
+                schemata.remove(wildcard_rule)
+                schemata.append(
+                    [wildcard_rule[0][:3] + "0" + wildcard_rule[0][4:], rule[1]]
+                )
+            elif (
+                rule[1] in ["1", 1]
+            ):  # if the output of the rule is 1, the corresponding maintenance rule output must also be 1
+                # modify the middle value of schemata(wildcard_rule) to "1"
+                schemata.remove(wildcard_rule)
+                schemata.append(
+                    [wildcard_rule[0][:3] + "1" + wildcard_rule[0][4:], rule[1]]
+                )
+            else:
+                raise ValueError("Rule output must be 0 or 1")
+    maintenance_rules = schemata
+    return maintenance_rules
+
+
+def fill_missing_outputs_as_maintenance(node: BooleanNode) -> BooleanNode:
+    """
+    Generate a node with missing output values filled as maintenance rules.
+
+    Args:
+        node (BooleanNode) : The node to generate with. The node must contain missing '?' output values.
+
+    Returns:
+        A BooleanNode object with missing output values filled as maintenance rules.
+    """
+
+    # check if there are any missing output values
+    if "?" not in node.outputs:
+        # raise ValueError("There are no missing output values in the node.")
+        return node # removed error to allow for complete functions to pass through smoothly
+    # Replace '?' in generated_node.outputs with 0 or 1 dependint on the middle element of the respective rule
+    lut = node.look_up_table().values.tolist()
+    middle_element = len(lut[0][0]) // 2  # middle input is reduced by 1 since indexing starts from 0
+    new_outputs = []
+    for lut_row in lut:
+        if lut_row[1] == "1":
+            new_outputs.append("1")
+        elif lut_row[1] == "0":
+            new_outputs.append("0")
+        else:
+            new_outputs.append("1" if lut_row[0][middle_element] == "1" else "0")
+    return BooleanNode.from_output_list(new_outputs)
+
+
+def filter_for_ke(generated_nodes, ke, original_ke, epsilon=0.01):
+    """
+    Filter for nodes that are within epsilon of the desired effective connectivity.
+
+    Args:
+    generated_nodes (iterator): List of generated nodes.
+    ke (float): Desired effective connectivity.
+    epsilon (float): Tolerance for effective connectivity.
+
+    Returns:
+    generated_node (BooleanNode): Node that is within epsilon of the desired effective connectivity.
+
+    """
+    generated_node = None
+    closest_node = None
+    closest_ke = None
```

Comment:
Are those initial settings of the parameters to "None" necessary? Because it might make debugging more difficult later.